modules/markdown/assets/js/components/markdown-live-preview.jsx

```diff
@@ -0,0 +1,164 @@
+	if ( this.props.onChange ) {
+		// Clone event with Object.assign to avoid
+		// "Cannot assign to read only property 'target' of object"
+		evt = Object.assign( {}, evt, {
+			target: {
+				value: source
+			}
+		} );
+		this.props.onChange( evt );
+	}
+};
+
+const sourceIsEmpty = function( source ) {
+	return ! source;
+};
+
+const ignoreLastInput = function( source ) {
+	const SPACE = String.fromCharCode( 32 );
+	const NO_BREAK_SPACE = String.fromCharCode( 160 );
+
+	return endsWith( source, SPACE ) || endsWith( source, NO_BREAK_SPACE );
+};
+
+const emptyState = '<p></p>';
+
+const renderMarkdownPreview = function( evt ) {
+	if ( ! this.htmlEl ) {
+		return true;
+	}
+
+	// We need to delete the last two new lines added by the browser for the
+	// last node in the component's content. This behaviour interferes with
+	// headings parsing.
+	const source = stripTrailingNewLines( this.htmlEl.innerText );
+
+	triggerOnChange.call( this, evt, source );
+
+	// if there's no source, we don't need to parse anything
+	if ( sourceIsEmpty( source ) ) {
+		this.htmlEl.innerHTML = emptyState;
+		return true;
+	}
+
+	// commonmark doesn't allow trailing spaces in paragraphs, so if we've
+	// added a space, calling markdownIt now will remove it
+	if ( ignoreLastInput( source ) ) {
+		return true;
+	}
+
+	this.setState( { restoreCaretPosition: null } );
+
+	if ( source ) {
+		const html = renderHTML( source );
+
+		this.setState( {
+			html
+		} );
+	}
+
+	return true;
+};
+
+const OBSERVER_CONFIG = { subtree: true, characterData: true };
+
+const isIE11 = !! window.MSInputMethodContext && !! document.documentMode;
+
+export default class MarkdownLivePreview extends React.Component {
+
+	constructor( props ) {
+		super();
+
+		const { source } = props;
+
+		this.state = {
+			html: source ? renderHTML( source ) : emptyState,
+		};
+	}
+
+	shouldComponentUpdate( nextProps, nextState ) {
+		if ( this.state.html !== nextState.html ) {
+			return true;
+		}
+	}
+
+	getSnapshotBeforeUpdate( prevProps, prevState ) {
+		if ( this.state.html !== prevState.html ) {
+			this.state.restoreCaretPosition = saveCaretPosition( this.htmlEl );
+		}
+	}
+
+	componentDidUpdate() {
+		// once the component has be rendered, we can restore the caret position
+		if ( this.state.restoreCaretPosition ) {
+			this.state.restoreCaretPosition();
+		}
+	}
+
+	componentDidMount() {
+		if ( this.props.isSelected ) {
+			this.htmlEl.focus();
+		}
+		// onInput doesn't work for content editable elements in Internet Explorer 11,
+		// but we can use a MutationObserver instead
+		if ( isIE11 ) {
+			this.observer = new MutationObserver( ( mutations ) => {
+				mutations.forEach( renderMarkdownPreview.bind( this ) );
+			} );
+			this.observer.observe( this.htmlEl, OBSERVER_CONFIG );
+		}
+	}
+
+	render() {
+		const { ...props } = this.props;
+
+		return createElement(
+			'div',
+			{
+				...props,
+				ref: ( e ) => this.htmlEl = e,
+				onBlur: renderMarkdownPreview.bind( this ),
```

Comment:
`bind` creates a new function on every render. The `shouldComponentUpdate` looks like it will make this less of a problem, but we should avoid doing this regardless as it can lead to unexpected behaviour.

Reply:
Yeah, you're totally right. Actually, I found another place where we're creating a new function quite often: [here](https://github.com/Automattic/jetpack/pull/9705/files/30c51379a59b6ccff6d49a2b5e71e267f529f704#diff-26458d622e729f313ea1b570125ab5adR36)
I'll fix both issues 👍